### PR TITLE
Deprecate methods prefixed with underscore (_) + replace with unprefixed methods

### PR DESCRIPTION
**Pull request by udidol**
Documentation has to be updated as well!
All mentions of the prefixed methods such as `_register_controls()` in the documentation should be changed to the unprefixed version (such as `register_controls()`.

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [X] Other... Please describe: **It is a refactor but the API does change**

## Summary

This PR can be summarized in the following changelog entry:

* Replaced control-related methods `_register_controls()`, `_get_initial_config()`, `_content_template()`, `_init()`, with unprefixed versions (e.g. `register_controls()`, `get_initial_config`, etc.)

## Description
An explanation of what is done in this PR

* Replaced control-related methods `_register_controls()`, `_get_initial_config()`, `_content_template()`, `_init()`, with unprefixed versions (e.g. `register_controls()`, `get_initial_config`, etc.).
* Replaced all of these methods' uses across the Elementor Core plugin

## Test instructions
This PR can be tested by following these steps:

* General Sanity tests, general testing of controls.

## Quality assurance

- [X] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: core/dynamic-tags/base-tag.php.rej [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
++		// TODO: This is for backwards compatibility starting from 2.9.0
++		// This if statement should be removed when the method is hard-deprecated, in version 3.3.0
++		if ( method_exists( $this, '_register_controls' ) ) {
++			$this->_register_controls();
++		} else {
++			$this->register_controls();
++		}
+ 
+ 		$this->end_controls_section();
+
```
Comment (kobizz) [inline core/dynamic-tags/base-tag.php.rej]: Removed this file

File: includes/base/controls-stack.php [hunk 2]
```diff
@@ -1641,7 +1647,7 @@ public function end_controls_tab() {
 	public function print_template() {
 		ob_start();
 
-		$this->_content_template();
+		$this->content_template();
```
Comment (kobizz) [inline includes/base/controls-stack.php]: Add a check if the old method exists

File: includes/base/controls-stack.php [hunk 3]
```diff
@@ -1796,17 +1802,35 @@ protected function get_init_settings() {
...
 	}
 
+	/**
+	 * Get initial config.
+	 *
+	 * Retrieve the current element initial configuration - controls list and
+	 * the tabs assigned to the control.
+	 *
+	 * @since 1.4.0
+	 * @deprecated 2.9.0 Implement `get_initial_config()` instead
```
Comment (kobizz) [inline includes/base/controls-stack.php]: Use "use" instead of "implement". Same for all occurrences

File: includes/base/controls-stack.php [hunk 4]
```diff
@@ -1877,25 +1916,54 @@ protected function _content_template() {}
 	protected function init_controls() {
 		Plugin::$instance->controls_manager->open_stack( $this );
 
-		$this->_register_controls();
+		// TODO: This is for backwards compatibility starting from 2.9.0
+		// This `if` statement should be removed when the method is removed
+		if ( method_exists( $this, '_register_controls' ) ) {
+			$this->_register_controls();
+		} else {
+			$this->_register_controls();
```
Comment (kobizz) [inline includes/base/controls-stack.php]:
```suggestion
			$this->register_controls();
```
Move it to version 3.0.0

File: includes/base/controls-stack.php [hunk 5]
```diff
@@ -1877,25 +1916,54 @@ protected function _content_template() {}
...
+	protected function _init( $data ) {
+		// _deprecated_function( __METHOD__, '2.9.0', 'init' );
+
+		// TODO: This is for backwards compatibility starting from 2.9.0
+		// This `if` statement should be removed when the method is removed
+		if ( method_exists( $this, '_init' ) ) {
+			$this->_init( $data );
+		} else {
+			$this->init( $data );
+		}
```
Comment (kobizz) [inline includes/base/controls-stack.php]: This check shouldn't be here

File: includes/base/controls-stack.php [hunk 6]
```diff
@@ -1961,7 +2029,7 @@ private function sanitize_settings( array $settings, array $controls = [] ) {
 	 */
 	public function __construct( array $data = [] ) {
 		if ( $data ) {
-			$this->_init( $data );
+			$this->init( $data );
```
Comment (kobizz) [inline includes/base/controls-stack.php]: Here you have to do the check

File: includes/base/controls-stack.php.rej [hunk 7]
```diff
@@ -0,0 +1,10 @@
+diff a/includes/base/controls-stack.php b/includes/base/controls-stack.php	(rejected hunks)
+@@ -1766,7 +1766,7 @@
+ 	 * @deprecated 2.9.0 Implement `register_controls()` instead
+ 	 * @access protected
+ 	 */
+-	protected function register_controls() {
++	protected function _register_controls() {
+ 		_deprecated_function( __METHOD__, '2.9.0', 'register_controls' );
+ 
+ 		$this->register_controls();
```
Comment (kobizz) [inline includes/base/controls-stack.php.rej]: Remove this file

File: includes/base/element-base.php [hunk 8]
```diff
@@ -867,7 +867,7 @@ protected function _print_content() {
 	 *
 	 * @return array The initial config.
 	 */
-	protected function _get_initial_config() {
+	protected function get_initial_config() {
```
Comment (kobizz) [inline includes/base/element-base.php]: Change `since` everywhere

**Review by kobizz — CHANGES_REQUESTED** [review]

File: includes/base/controls-stack.php [hunk 9]
```diff
@@ -1641,7 +1647,11 @@ public function end_controls_tab() {
 	public function print_template() {
 		ob_start();
 
-		$this->_content_template();
+		if ( method_exists( $this, '_content_template' ) ) {
```
Comment (kobizz) [inline includes/base/controls-stack.php]: Add "TODO" note

File: includes/controls/base.php [hunk 10]
```diff
@@ -94,7 +94,7 @@ public function enqueue() {}
 	 *
 	 * Note that the content template is wrapped by Base_Control::print_template().
 	 *
-	 * @since 1.5.0
+	 * @since 2.9.0
```
Comment (kobizz) [inline includes/controls/base.php]: Reverse the change

File: includes/controls/base.php [hunk 11]
```diff
@@ -114,7 +114,15 @@ public function enqueue() {}
 		?>
 		<script type="text/html" id="tmpl-elementor-control-<?php echo esc_attr( $this->get_type() ); ?>-content">
 			<div class="elementor-control-content">
-				<?php $this->content_template(); ?>
+				<?php
```
Comment (kobizz) [inline includes/controls/base.php]: Reverse

File: includes/controls/code.php [hunk 12]
```diff
@@ -55,7 +55,7 @@ protected function get_default_settings() {
 	 * template. The variables for the class are available using `data` JS
 	 * object.
 	 *
-	 * @since 1.0.0
+	 * @since 2.9.0
```
Comment (kobizz) [inline includes/controls/code.php]: Reverse in all controls

**Review by kobizz — CHANGES_REQUESTED** [review]

**Review by kobizz — APPROVED** [review]

**Comment by eatbuildplay** [comment]
This is confusing to me as a first time widget builder, I tried using the unprefixed register_controls() and it doesn't work, so I'm using _register_controls() which does work and I checked the Elementor core widgets, they are also using _register_controls().

**Comment by udidol** [comment]
> This is confusing to me as a first time widget builder, I tried using the unprefixed register_controls() and it doesn't work, so I'm using _register_controls() which does work and I checked the Elementor core widgets, they are also using _register_controls().

Hi @eatbuildplay,
Some of the code in this PR was later modified. `_register_controls()` will be staying prefixed for now. We will issue a blog post in [our developers blog](https://developers.elementor.com/blog/) that will address this in due time.

For more info on building widgets, consult with our developer documentation: [https://developers.elementor.com/creating-a-new-widget/](https://developers.elementor.com/creating-a-new-widget/)

**Pull request merged**